ConfigSpace/util.pyx

```diff
@@ -436,3 +436,185 @@ def fix_types(configuration: dict,
+            if param.log:
+                lower, upper = np.log([param.lower, param.upper])
+            else:
+                lower, upper = param.lower, param.upper
+
+            if param.name in num_steps_dict:
+                num_steps = num_steps_dict[param.name]
+                grid_points = np.linspace(lower, upper, num_steps)
+            else:
+                grid_points = np.arange(lower, upper + param.q, param.q) # check for log and for rounding issues
+
+            if param.log:
+                grid_points = np.exp(grid_points)
+            grid_points = grid_points.astype(int)
+
+            # Avoiding rounding off issues
+            if grid_points[0] < param.lower:
+                grid_points[0] = param.lower
+            if grid_points[-1] > param.upper:
+                grid_points[-1] = param.upper
+
+            return tuple(grid_points)
+
+        else:
+            raise TypeError("Unknown hyperparameter type %s" % type(param))
+
+    def get_cartesian_product(value_sets, hp_names):
+        '''
+        Returns a grid for a subspace of the configuration with given hyperparameters and their grid values.
+
+        Takes a list of tuples of grid values of the hyperparameters and list of hyperparameter names. The outer list iterates over the hyperparameters corresponding to the order in the list of hyperparameter names. The inner tuples contain grid values of the hyperparameters for each hyperparameter.
+
+        Parameters
+        ----------
+        value_sets: list of tuples
+            Same description as return value of get_value_set()
+
+        hp_names: list of strs
+            List of hyperparameter names
+
+        Returns
+        -------
+        list of dicts
+            List of configuration dicts
+
+        '''
+        grid = []
+        import itertools
+        for i, element in enumerate(itertools.product(*value_sets)):
+            config_dict = {}
+            for j, hp_name in enumerate(hp_names):
+                config_dict[hp_name] = element[j]
+            grid.append(config_dict)
+
+        return grid
+
+
+    value_sets = [] # list of tuples: each tuple within is the grid values to be taken on by a Hyperparameter
+    hp_names = []
+
+    # Get HP names and allowed grid values they can take for the HPs at the top level of ConfigSpace tree
+    for hp_name in configuration_space._children['__HPOlib_configuration_space_root__']:
+        value_sets.append(get_value_set(num_steps_dict, hp_name))
+        hp_names.append(hp_name)
+
+    # Create a Cartesian product of above allowed values for the HPs. Hold them in an "unchecked" list because some of the conditionally dependent HPs may become active for some of the elements of the Cartesian product and in these cases creating a Configuration would throw an Error (see below).
+    unchecked_grid_pts = get_cartesian_product(value_sets, hp_names) # Creates a list of Configuration dicts

```

Comment:
Could you please use a double ended queue (`collections.deque`) here to make the code a bit more intuitive?

Reply:
Done